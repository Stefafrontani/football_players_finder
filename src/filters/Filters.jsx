import React from 'react';
import { useDispatch } from 'react-redux';
import { setFilters } from './filtersAction';
import './filtersStyles/Filters.css';

const Filter = () => {
  const dispatch = useDispatch();

  const handleFieldError = {
    name: handleNameField,
    age: handleAgeField,
  };

  function checkName(value, pattern) {
    const matchPattern = value.match(pattern);
    return {
      isValidName: Boolean(!matchPattern || matchPattern[0] !== matchPattern.input),
      matchPattern
    }
  }

  function handleNameField(event) {
    const target = event.target;
    const value = target.value;
    const classList = target.classList;
    const attributes = target.attributes;
    const patternValue = attributes.pattern.value;

    const { isValidName, matchPattern } = checkName(value, patternValue);

    isValidName ? classList.add('wrong-value') : classList.remove('wrong-value');
    event.target.value = isValidName
      ? matchPattern
        ? matchPattern[0]
        : ''
      : value;
  }

  function handleAgeField(event) {
    const target = event.target;
    const classList = target.classList;

    const isValid = target.validity.valid;

    isValid ? classList.remove('wrong-value') : classList.add('wrong-value');
  }

  function onChange(event) {
    handleFieldError[event.target.name] && handleFieldError[event.target.name](event);
    hasValue(event);
  }

  function hasValue(event) {
    const target = event.target;
    const targetValue = target.value;
    const classList = target.classList;

    targetValue ? classList.add('has-value') : classList.remove('has-value');
  }

  function onSubmit(event) {
    event.preventDefault();
    event.stopPropagation();

    const form = event.target;
    const formData = new FormData(form);

    let filters = {}

    for (let [key, value] of formData) {
      filters[key] = value;
    }

    dispatch(setFilters(filters));
  }

  return (
    <div className="mb16" style={{ width: '100%' }}>
      <form onSubmit={onSubmit} className="flex flex-row flex-items-center flex-content-center">
        <div className="mr16 field-container position-relative">
          <input onChange={onChange} id="name" type="text" pattern="[A-Za-z]([A-Za-z]|[' '](?=[A-Za-z]))+|[A-Za-z]" name="name"/>
          <span/>
          <label className="position-absolute fs15" htmlFor="name"> Name </label>
        </div>
        <div className="mr16 field-container position-relative">
          <input onChange={onChange} id="position" type="text" name="position"/>
          <span/>
          <label className="position-absolute fs15" htmlFor="position"> Position </label>
        </div>
        <div className="mr16 field-container position-relative">
          <input onChange={onChange} id="age" type="number" min="18" max="40" name="age"/>
          <span/>
          <label className="position-absolute fs15" htmlFor="age"> Age </label>
        </div>
          <input type="submit"/>
      </form>
    </div>
  );
}

export default Filter;