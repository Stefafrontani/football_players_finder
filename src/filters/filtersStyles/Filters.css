:root {
  --form-transition-time: 0.2s;
}

input:not([type="submit"]) {
  padding: 3px 5px; /* same as input ~ label left (*) */
  border-bottom: 1px solid rgb(75, 75, 230, 0.2);
}

input[type="submit"] {
  background-color: transparent;
  border-radius: 10px;
  box-shadow: 2px 2px 4px 2px rgba(0, 0, 0, 0.2);
  padding: 5px 10px;
}

/* Place label absolute  */
input ~ label {
  top: 50%;
  left: 3px; /* (*) */
  transform: translate(0, -50%);
  transition: transform var(--form-transition-time), top var(--form-transition-time), left var(--form-transition-time), color var(--form-transition-time);
}

/* Move label when input focus  */
input:focus ~ label,
input.has-value ~ label {
  top: -60%;
  left: 0px;
  transform: scale(0.8);
  transition: transform var(--form-transition-time), top var(--form-transition-time), left var(--form-transition-time), color var(--form-transition-time);
  color: rgb(75, 75, 230);
}

/* Fake border bottom growth */
.field-container {} /*  */
.field-container input ~ span {
  position: absolute;
  content: "";
  top: 100%;
  left: 0%;
  height: 1px;
  width: 0px;
  background-color: rgb(75, 75, 230);

}
.field-container input:focus ~ span,
.field-container input.has-value ~ span {
  width: 100%;
  transition: width var(--form-transition-time);
}

input.wrong-value ~ span {
  background-color: rgb(185, 76, 76);
}

input[name="name"].wrong-value ~ span:after {
  content: "Please, only text"
}
input[name="age"].wrong-value ~ span:after {
  content: "Please, only numbers between 18-40";
  width: 200px;
}

input.wrong-value ~ span:after {
  width: 100%;
  position: absolute;
  text-align:left;
  top: 30%;
  margin-top: 4px;
  font-size: 10px;
  color: rgb(185, 76, 76);
  left: 0;
}