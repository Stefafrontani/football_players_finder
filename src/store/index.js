import { createStore, applyMiddleware } from 'redux';
import { logger } from 'redux-logger';
import thunk from 'redux-thunk';
import appReducer from '../app/appReducer';

const store = createStore(appReducer, applyMiddleware(thunk, logger));

export default store;