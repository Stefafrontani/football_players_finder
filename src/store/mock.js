import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

const mockApplicationState = {
    filters: {},
    playersList: []
}

const mockServiceCreator = (body, succeeded = true) => () =>
    new Promise((resolve, reject) => {
        setTimeout(() => succeeded ? resolve(body) : reject(body), 100)
    })

const getMockProvider = (partialState) => {
    const mockStore = configureMockStore();
    const stateToApply = partialState ? { ...mockApplicationState, ...partialState } : mockApplicationState;
    const store = mockStore(stateToApply);

    return {
        MockProvider: ({ children }) => {
            return <Provider store={store}>{children}</Provider>
        }
    };
};

const initializeMockProvider = (partialState) => {
    const { MockProvider } = getMockProvider(partialState);

    return {
      MockProvider
    }
  }

export { mockStore, getMockProvider, initializeMockProvider, mockServiceCreator };