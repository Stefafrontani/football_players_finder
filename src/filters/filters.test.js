import React from 'react';
import Filters from './Filters';
import { initializeMockProvider } from '../store/mock';

describe('<Filters />', () => {
    let wrappedComponent;
    beforeAll(() => {
        const { MockProvider } = initializeMockProvider({});
        wrappedComponent = <MockProvider><Filters /></MockProvider>
    });
    it('Snapasdhot', () => {
        const component = renderer.create(wrappedComponent);
        const tree = component.toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('', () => {
        const wrapper = shallow(wrappedComponent);
        expect(true).toBe(true);
    });
});