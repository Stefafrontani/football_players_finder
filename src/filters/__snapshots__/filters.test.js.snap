// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Filters /> Snapasdhot 1`] = `
<div
  className="mb16"
  style={
    Object {
      "width": "100%",
    }
  }
>
  <form
    className="flex flex-row flex-items-center flex-content-center"
    onSubmit={[Function]}
  >
    <div
      className="mr16 field-container position-relative"
    >
      <input
        id="name"
        name="name"
        onChange={[Function]}
        pattern="[A-Za-z]([A-Za-z]|[' '](?=[A-Za-z]))+|[A-Za-z]"
        type="text"
      />
      <span />
      <label
        className="position-absolute fs15"
        htmlFor="name"
      >
         Name 
      </label>
    </div>
    <div
      className="mr16 field-container position-relative"
    >
      <input
        id="position"
        name="position"
        onChange={[Function]}
        type="text"
      />
      <span />
      <label
        className="position-absolute fs15"
        htmlFor="position"
      >
         Position 
      </label>
    </div>
    <div
      className="mr16 field-container position-relative"
    >
      <input
        id="age"
        max="40"
        min="18"
        name="age"
        onChange={[Function]}
        type="number"
      />
      <span />
      <label
        className="position-absolute fs15"
        htmlFor="age"
      >
         Age 
      </label>
    </div>
    <input
      type="submit"
    />
  </form>
</div>
`;
